### PR TITLE
[shardformer] bert support sequence parallel.

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4454

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

bert support sequence parallel.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/_operation.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ class _SplitForwardGatherBackward(torch.autograd.Function):
     def forward(ctx, input_, dim, process_group):
         ctx.process_group = process_group
         ctx.dim = dim
-        return _split(input_, dim, process_group)
+        return _split(input_, dim, process_group).clone()
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:370]: Why clone?

File: colossalai/shardformer/modeling/bert.py [hunk 2]
```diff
@@ -29,13 +29,25 @@
...
 
 class BertPipelineForwards:
     '''
     This class serves as a micro library for forward function substitution of Bert models
     under pipeline setting.
     '''
+    tp_shard_config: ShardConfig = None
+
+    @staticmethod
+    def set_shard_config(cls, shard_config):
```
Comment (ver217) [inline colossalai/shardformer/modeling/bert.py:44]: You'd better inject this when replace forward function. This Class is used like namespace.

Comment (flybird11111) [reply colossalai/shardformer/modeling/bert.py]: > You'd better inject this when replace forward function. This Class is used like namespace.

updated

File: colossalai/shardformer/modeling/bert.py [hunk 3]
```diff
@@ -56,6 +58,7 @@ def bert_model_forward(
         stage_manager: Optional[PipelineStageManager] = None,
         hidden_states: Optional[torch.FloatTensor] = None,    # this is from the previous stage
         stage_index: Optional[List[int]] = None,
+        tp_shard_config: ShardConfig = None,
```
Comment (ver217) [inline colossalai/shardformer/modeling/bert.py:61]: It should be `shard_config`

File: colossalai/shardformer/modeling/bert.py [hunk 4]
```diff
@@ -177,6 +180,14 @@ def bert_model_forward(
         start_idx, end_idx = stage_index[0], stage_index[1]
         # layer_outputs
         layer_outputs = hidden_states if hidden_states is not None else None
+
+        # split the input tensor along sequence dimension
+        # [batch_size, seq_len, hidden_size] -> [batch_size, seq_len/TP_size, hidden_size]
```
Comment (ver217) [inline colossalai/shardformer/modeling/bert.py:185]: What if seq len is not dividable by tp size?

File: colossalai/shardformer/layer/_operation.py (hunk 1 above)
Comment (flybird11111) [reply colossalai/shardformer/layer/_operation.py]: > Why clone?

updated.

File: colossalai/shardformer/modeling/bert.py (hunk 4 above)
Comment (flybird11111) [reply colossalai/shardformer/modeling/bert.py:185]: Inside the _split function, there is an assert for dim_size // world_size.

Comment (flybird11111) [reply colossalai/shardformer/modeling/bert.py:185]: > What if seq len is not dividable by tp size?

Inside the _split function, there is an assert for dim_size // world_size.

**Review by ver217 — APPROVED** [review]

**Pull request merged**